site/content/features/sealedtypes.md

```diff
@@ -0,0 +1,506 @@
+---
+title: Sealed Classes (JEP 360)
+copyright: Cay S. Horstmann 2020. All rights reserved.
+jep: 360
+jdkversion: 15
+type: "sandbox"
+---
+
+
+Java 15 provides a preview of “sealed types”—types with a fixed set of direct subtypes. This feature allows for accurate modeling of type hierarchies that should not be open to arbitrary inheritance, and it allows the compiler to check for exhaustive pattern matching.
+
+## Controlling Subtypes
+
+Unless a class is declared `final`, anyone can form a subclass of it. What if you want to have more control? For example, suppose you feel the need to write your own JSON library because none of the ones out there do exactly what you need.
+
+The JSON standard says that a JSON value is an object, array, string, number, Boolean, or `null`. The Java language had, until now, no way of expressing that a `JSONValue` should be exactly one of those six types.
+
+Java 15 provides a preview of “sealed types”, where you get just that control. You can define `JSONValue` to be `sealed`, and list the subclasses in a `permits` clause:
+
+```
+public sealed abstract class JSONValue
+   permits JSONObject, JSONArray, JSONString, JSONNumber, JSONBoolean, JSONNull {
+   . . .
+}
+```
+
+If anyone were to form another subclass, such as
+
+```
+public class JSONComment extends JSONValue { . . . } // Error
+```
+
+this would be an error. And that's just as well, since JSON doesn't allow for comments. As you can see, sealed classes allow for accurate modeling of domain constraints. 
+
+The dictionary defines sealing as (1) affixing a mark that attests to quality or absence of tampering, or (2) securing against access or damage. I am not sure that either of these describes what the `sealed` keyword does. A sealed class is protected from one specific evil, namely promiscuous subclassing.
+
+```
+public sealed abstract class JSONValue
+      permits JSONObject, JSONArray, JSONString, JSONNumber, JSONBoolean, JSONNull {
+   // . . .
+}
+```
+
+```
+public final class JSONObject extends JSONValue {
+   // . . .
+}
+```
+
+```
+public final class JSONArray extends JSONValue {
+   // . . .
+}
+```
+
+```
+public final class JSONString extends JSONValue {
+   // . . .
+}
+```
+
+```
+public final class JSONNumber extends JSONValue {
+   // . . .
+}
+```
+
+```
+public final class JSONBoolean extends JSONValue {
+   // . . .
+}
+```
+
+```
+public final class JSONNull extends JSONValue {
+   // . . .
+
+}
+```
+
+```
+public class Sandbox {
```

Comment:
I wonder whether these source snippets should have gone into a sandbox tag?

Reply:
Yes, my bad. Something went wrong in the conversion from XHTML to Markdown. I just fixed that.